Use rapidjson in loki

xlqian opened the pull request.

xlqian commented:
sorry, we'll reopen it when more tests are added

kevinkreiser commented:
no worries! dont forget that when making makefile changes that you must use tabs for the whitespace at the beginning of a line.

kevinkreiser commented:
also it seems we had forgotton to make prs build from forks in our continuous integration so it didnt run the tests. feel free to reopen and we can make it run the tests. or if you want to write more tests feel free to do that as well. we'd like to add back some optional logic to loki that handles disconnected portions of the network so it'd be great to get your changes in before that.

kevinkreiser commented on Makefile.am:
```diff
@@ -98,6 +98,7 @@ nobase_include_HEADERS = \
 	valhalla/baldr/nodeinfo.h \
 	valhalla/baldr/location.h \
 	valhalla/baldr/pathlocation.h \
+        valhalla/baldr/rapidjson_utils.h \
```
dont forget makefiles must have only tabs in them!

kevinkreiser commented on src/loki/isochrone_action.cc:
```diff
@@ -1,7 +1,9 @@
 #include "loki/service.h"
 #include "loki/search.h"
 #include "baldr/datetime.h"
+#include "/baldr/rapidjson_utils.h"
```
extra slash here

kevinkreiser commented on src/baldr/location.cc:
```diff
@@ -4,10 +4,14 @@
 #include "midgard/logging.h"
 #include "midgard/util.h"
 
+#include <rapidjson/document.h>
```
maybe common includes (like these?) could just be included in `baldr/rapidjson_utils.h` ?

kevinkreiser commented on src/loki/locate_action.cc:
```diff
@@ -3,8 +3,11 @@
+#include "baldr/rapidjson_utils.h"
 #include "midgard/logging.h"
 
++#include <rapidjson/pointer.h>
```
looks like  a typo

xlqian commented:
@kevinkreiser Actu

xlqian commented:
@kevinkreiser  Thank you for your quick feedback. I assume that I opened this PR by accident and codes weren't ready to be reviewed yet. (because we worked in baldr, sif and loki in their own repository and we had to move all our modifications here due to the unification of valhalla. and I intended to open a PR for an internal review) 

We haven't finished adding tests, but all tests pass on our machine. Now we reopen the PR so that, if you wish, you can do more tests. 

Note:
* **We added `-DRAPIDJSON_HAS_STDSTRING` in CXXFLAGS**, codes won't compile if this is missing  
* We updated the version of Rapidjson, which allows us to do `Is<T>` and `Get<T>` like this: https://github.com/valhalla/valhalla/pull/502/files#diff-a23d400977b64fbc2fd725709019a6cbR15

Have a nice day

Navitia Team

TeXitoi commented on src/baldr/location.cc:
```diff
@@ -46,6 +47,40 @@ boost::property_tree::ptree Location::ToPtree() const {
+    location.AddMember(rapidjson::Value{}.SetString(attr.c_str(), attr.length(), a),
+            rapidjson::Value{}.SetString(s.c_str(), s.length(), a),
+            a);
+  };
```
Can't we just use `std::string` version?

TeXitoi commented on src/loki/service.cc:
```diff
@@ -164,10 +166,14 @@ namespace valhalla {
-      std::string method_options = "costing_options." + *costing;
-      auto costing_options = request.get_child(method_options,{});
+      std::string method_options = "/costing_options/" + *costing;
+      auto* method_options_ptr = rapidjson::Pointer(method_options.c_str()).Get(request);
```
directly use std::string

TeXitoi commented on test/location.cc:
```diff
@@ -101,102 +101,112 @@ void test_from_json() {
+      Location loc = Location::FromJson("something", m);
+      throw std::runtime_error("This should have been malformed json");
+    } catch (...) {
+    }
```
@kevinkreiser  Why this do something that will just be trashed whatever, then throw, and all that is catched nothing? That's just useless code, no?

kevinkreiser replied:
the test fails if there is an uncaught exception. this bit of code is trying to make sure that bad json will throw an exception. to do that properly one should probably do something like:

```c++11
for(const auto& m : methods) {
  try {
    Location loc = Location::FromJson("something", m);
  } //this should catch something
  catch (...) {
    continue;
  }
  //if we got here then the above didnt throw and that is wrong
  throw std::runtime_error("This should have been malformed json");
}
```

kevinkreiser commented:
hey @xlqian and @TeXitoi i havent had a chance to review this fully yet but i will do so today! thanks very much for all of your work!

kevinkreiser commented:
sorry about this its taken me a long time to review this i hope to be finished with it early next week. sorry for the delay!

xlqian commented:
No problem!

xlqian commented:
Please don't forget our PR when you've got time to spend on it, thanks.

kevinkreiser commented:
@xlqian i havent forgotten about it just havent had the time yet. also we have been thinking about using protobuf to pass around the request object instead of constantly parsing json. apparently protobuf can parse json directly these days although the docs say its 'not heavily optimized'. Perhaps we can  use rapidjson to parse and then throw it into a protobuf object in loki then the rest of the system could use protobuf throughout. we'll have to see! at any rate ill try to fix the conflicts and look at this again today.

xlqian commented:
@kevinkreiser Hi, the protobuf has also been mentioned by @TeXitoi during one of our internal discussions actually. It's a very good idea! We have been thinking of making a mini customized service which would communicate with loki and thor through protobuf. As you may have noticed, this will require a lot of work, much more than using rapidjson. Let's say using rapidjson is a quick cheap solution to improve slightly the perfomance and protobuf, IMHO, would be the final ultimate solution.

kevinkreiser commented:
i have found a few regressions with this and am working on fixing them.

xlqian commented:
@kevinkreiser  Thx!

kevinkreiser commented:
ok so there were a bunch of issues here:

* isochrones didnt work because it was using the wrong node of the request object
* many_to_one didnt work because it was setting the targets using `.End()` which is invalid
* route didnt work because it had a leading `/` when accessing the locations part of the DOM
* locate didnt work correctly because the verbose boolean flag `is_convertable` to `double` but is `!rapidjson::IsNumeric()`
* tests were broken for kvp that had empty strings in it essentially

given that this touches every one of our systems, i think we need a very thorough review/testing phase of this before we merge it. it seems to be working at a bare minimum now but its hard to test all of the permutations!

@xlqian @TeXitoi @dnesbitt61 @kdiluca @dgearhart

xlqian commented:
Hi, Thank you for your response and we appreciate it! In terms of regression(sorry for that), could you tell us how we can help to fix the regression? could you please give us some failed test cases so that we can work together? @TeXitoi

kevinkreiser commented:
@xlqian @TeXitoi i've already fixed the regressions i've found. im not sure how github does it but maybe the commits are in your branch, they show up here in the PR commits. I was simply saying that with such a large change and after finding so many issues that it would be wise for us to do more extensive testing before merging and releasing this. @dgearhart @dnesbitt61 what do you think?

dnesbitt61 commented:
I agree that we should do some more extensive testing prior to release. Will try to schedule that for late this week so we can merge (assuming all tests go well) early next week.

TeXitoi commented:
Thanks!
